### PR TITLE
Refactor current baseline configs

### PR DESCRIPTION
**Pull request by ludziej**
# Description
This PR is meant to reduce amount of redundant information in our configs, especially by making argparse names consistent with our current terminology that is invariant of model size and by expanding our config tree (and fixing configs that were not runnable). Moreover, to simplify grid running, you can now pass multiple configs to run_exp_remotely.

Now for example the inherinatance goes like this: Granular_mini -> expert_choice_mini -> expert_choice_common -> gpt_common -> common. Each of this files has only several lines, that clearly relate to its name.

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [x] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [x] Have you successfully executed the linter and tests?
   - [x] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [x] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [x] Are all functions placed in appropriate files?
   - [x] Is the PR name meaningful and descriptive?
   - [x] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: research/conditional/utils/argparse.py [hunk 1]
```diff
@@ -137,6 +137,9 @@ def introduce_parser_arguments(
         help="This argument is deprecated. Provide either (total_experts_width, n_experts, effective_dff) or (expert_size, n_experts, topk_fraction) instead.",
     )
     parser.add_argument("--total_experts_width", type=int)
+    parser.add_argument("--granularity", type=int)
+    parser.add_argument("--expansion_rate", type=int)
+    parser.add_argument("--effective_dff_x", type=int)
```
Comment (crewtool) [inline research/conditional/utils/argparse.py:142]: Could you provide explanations in 'help'?

File: research/conditional/train/configs/experiments/small_group.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  n_steps: 25_000
+  lr_warmup_steps: 2500
+
+  should_evaluate_dynamic_groupsize: true
+  n_eval_batches: 10
+  eval_interval: 10
+  ^simulate_group_size: [8]
+
+  # DEBUG TODO REMOVE
+  dataset_type: wikibook  # TODO TYLKO TO JEST TEST REMOVE
```
Comment (crewtool) [inline research/conditional/train/configs/experiments/small_group.yaml:19]: debug leftovers

File: research/conditional/train/configs/baselines/common.yaml [hunk 3]
```diff
@@ -20,10 +21,10 @@ params:
   # shallower bass
   cutoff: 256
   batch_size: 256
-  n_steps: 275_000
+  n_steps: 250_000
```
Comment (crewtool) [inline research/conditional/train/configs/baselines/common.yaml:24]: Why change?

File: lizrd/scripts/run_exp_remotely.sh [hunk 4]
```diff
@@ -34,5 +34,9 @@ run_grid_remotely() {
   ssh -t $host "$script"
 }
 
-# run your bash function
-run_grid_remotely $1 $2
+for i in "${@:2}"
```
Comment (crewtool) [inline lizrd/scripts/run_exp_remotely.sh:37]: Can you add a comment what is happening here, as none of us is an expert in bash ^^

File: research/conditional/train/configs/baselines/common.yaml [hunk 5]
```diff
@@ -1,17 +1,18 @@
-time: "2-00:00:00"
+runner: "research.conditional.train.cc_train"
+time: 1-00:00:00
```
Comment (crewtool) [inline research/conditional/train/configs/baselines/common.yaml:2]: Why change?

File: research/conditional/utils/model_utils.py [hunk 6]
```diff
@@ -207,12 +207,31 @@ def get_expert_choice_args(args):
...
         raise AssertionError(
-            "You must specify either total_experts_width, effective_dff, and n_experts or expert_size, topk_fraction, and n_experts"
+            "You must specify either total_experts_width, effective_dff, and n_experts "
+            "or expert_size, topk_fraction, and n_experts "
+            "or granularity, expansion_rate, and effective_dff_x "
         )
 
+    if set_arguments_option3:
+        # 4 is the standard dff_x, we assume it's defined relative to that
+        dff_x = 4
```
Comment (crewtool) [inline research/conditional/utils/model_utils.py:230]: This variable is a mystery to me. If it comes from the paper, maybe we can add a comment what it does above. If not, could we rename it?

**Review by crewtool — COMMENTED** [review]
Left comments

**Review by Simontwice — CHANGES_REQUESTED** [review]
remove run_exp_remotely.sh changes

**Review by Simontwice — APPROVED** [review]
ok

**Review by Simontwice — CHANGES_REQUESTED** [review]
supply a script that will replace all the hashes in my local configs with the current ones

File: research/conditional/train/configs/baselines/common.yaml (hunk 3 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:24]: +1

Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:24]: revert

File: research/conditional/train/configs/baselines/common.yaml [hunk 7]
```diff
@@ -20,10 +21,10 @@ params:
   # shallower bass
   cutoff: 256
   batch_size: 256
-  n_steps: 275_000
+  n_steps: 250_000
   scheduler: cosine
+  learning_rate: 1e-4
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/common.yaml:26]: this should never be here

File: research/conditional/train/configs/baselines/common.yaml [hunk 8]
```diff
@@ -20,10 +21,10 @@ params:
...
   lr_warmup_steps: 2500
-  final_lr_step: 250_000
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/common.yaml:26]: revert

File: research/conditional/train/configs/baselines/common.yaml [hunk 9]
```diff
@@ -1,17 +1,18 @@
-time: "2-00:00:00"
+runner: "research.conditional.train.cc_train"
+time: 1-00:00:00
 n_gpus: 1
 interactive_debug: false
-runner: "research.conditional.train.cc_train"
+name: "baseline"
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/common.yaml:5]: the code is not supposed to support names outside "params"

File: research/conditional/train/configs/baselines/common.yaml (hunk 5 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:2]: +1, revert

File: research/conditional/train/configs/baselines/common.yaml (hunk 7 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:26]: revert

File: lizrd/train/scheduler.py [hunk 10]
```diff
@@ -10,6 +10,8 @@ def get_scheduler(args):
             lr_warmup_steps=args.lr_warmup_steps, lr=args.learning_rate
         )
     elif args.scheduler == "cosine":
+        if args.final_lr_step is None:
+            args.final_lr_step = args.n_steps
```
Comment (Simontwice) [inline lizrd/train/scheduler.py:14]: I do not agree with this code, it essentially introduces a default parameter for `final_lr_step`

File: lizrd/scripts/grid.py [hunk 11]
```diff
@@ -46,6 +46,8 @@
         config["params"]["git_branch"] = args.git_branch
         config["params"]["path_to_entry_config"] = args.config_path
         config["params"]["all_config_paths"] = ",".join(all_config_paths)
+        if config["params"].get("name") is None:
+            config["params"]["name"] = config["name"]
```
Comment (Simontwice) [inline lizrd/scripts/grid.py:50]: revert, names should only be supplied in the "params" dict within configs.

**Review by Simontwice — CHANGES_REQUESTED** [review]
Baseline configs should stay the same (I support rearranging of arguments so it's consistent with other configs @ludziej, but some arguments that were added in `baselines/common.yaml` are unacceptable

**Comment by ludziej** [comment]
applied PR comments, also added lizrd/scripts/list_md5sum.sh script that calculates all md5sums of our config yamls

File: research/conditional/train/configs/baselines/common.yaml (hunk 7 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:27]: do not resolve, definitely revert

File: research/conditional/train/configs/baselines/common.yaml (hunk 8 above)
Comment (Simontwice) [reply research/conditional/train/configs/baselines/common.yaml:26]: why was this resolved. The final lr step is supposed to be 250K. the final training step is 275K. This line is required.

File: lizrd/scripts/list_md5sum.sh [hunk 12]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
+
+find $(dirname "$0")/../../research/conditional/train/configs/ -name '*.yaml' -type f -exec md5sum {} +
```
Comment (Simontwice) [inline lizrd/scripts/list_md5sum.sh:3]: nice! thanks

File: research/conditional/train/configs/baselines/gpt/dense/base.yaml [hunk 13]
```diff
@@ -0,0 +1,5 @@
+parent: research/conditional/train/configs/baselines/gpt/base.yaml
+md5_parent_hash: 94cd964259d1065e3ed2e26028f4865a
+
+params:
+  learning_rate: 1e-4
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/base.yaml:5]: revert, as discussed

File: research/conditional/train/configs/baselines/gpt/dense/medium.yaml [hunk 14]
```diff
@@ -0,0 +1,5 @@
+parent: research/conditional/train/configs/baselines/gpt/medium.yaml
+md5_parent_hash: 94cd964259d1065e3ed2e26028f4865a
+
+params:
+  learning_rate: 1e-4
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/medium.yaml:5]: revert, as discussed

File: research/conditional/train/configs/baselines/gpt/dense/mini.yaml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+parent: research/conditional/train/configs/baselines/gpt/mini.yaml
+md5_parent_hash: 94cd964259d1065e3ed2e26028f4865a
+
+params:
+  learning_rate: 1e-4
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/mini.yaml:5]: revert, as discussed

File: research/conditional/train/configs/baselines/gpt/dense/small.yaml [hunk 16]
```diff
@@ -0,0 +1,5 @@
+parent: research/conditional/train/configs/baselines/gpt/small.yaml
+md5_parent_hash: 94cd964259d1065e3ed2e26028f4865a
+
+params:
+  learning_rate: 1e-4
```
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/small.yaml:5]: revert, as discussed

File: research/conditional/train/configs/experiments/medium.yaml
Comment (Simontwice) [inline research/conditional/train/configs/experiments/medium.yaml:1]: this file cannot be in the toplevel of directory `experiments`. Experiments should be divided into subdirectories, similar to `configs/baselines`. I would like to see directories `contmoe`,`expert_choice (or granularity)`, `misc`. 
I would like this experiments to go into `misc`.

File: research/conditional/train/configs/baselines/gpt/granularity/4/base.yaml
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/granularity/4/base.yaml:1]: what does the name `4` name signify? It is not descriptive.

File: research/conditional/train/configs/baselines/gpt/dense/base.yaml
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/base.yaml:1]: If this file should be created, it should go to `configs/baselines/gpt/expert_choice/dense`.

File: research/conditional/train/configs/baselines/gpt/dense/medium.yaml
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/medium.yaml:1]: If this file should be created, it should go to `configs/baselines/gpt/expert_choice/dense`.

File: research/conditional/train/configs/baselines/gpt/dense/mini.yaml
Comment (Simontwice) [inline research/conditional/train/configs/baselines/gpt/dense/mini.yaml:1]: If this file should be created, it should go to `configs/baselines/gpt/expert_choice/dense`.

**Review by Simontwice — CHANGES_REQUESTED** [review]
A few details left, some are my personal request w.r.t. the quality of our config structure:

1. I can only speak for the common part of the config structure, and there I did as good a job as I was able to. I cannot, however, give good feedback on the `expert_choice`/`granularity` configs, which is why I want @kuba-krj to accept this code as well. This way I can be sure that all of this code is up to standard.
2. [my request] Please move files `mini`,`small`,`medium`,`base` contained in `research/conditional/train/configs/baselines/gpt/*`, to a new created directory `research/conditional/train/configs/baselines/gpt/dense`. This will clear up the structure of our configs and make everything much more readable.
3. Address comments I left in this round of review.

I really appreciate this effort and the mutual feedback we've been exchanging.

**Review by kuba-krj — APPROVED** [review]
k

**Review by Simontwice — CHANGES_REQUESTED** [review]
All good! Please update the MD5 hashes and we're done! Thanks

File: research/conditional/utils/argparse.py [hunk 17]
```diff
@@ -11,7 +11,9 @@ def introduce_parser_arguments(
     parser.add_argument("--ff_mode", type=str, default="vanilla")
     parser.add_argument("--n_blocks", type=int, required=True)
     parser.add_argument("--dmodel", type=int, required=True)
-    parser.add_argument("--dff", type=int, required=True)
+    parser.add_argument(
+        "--dff", type=int, required=False
+    )  # not used by MoT & granularity
```
Comment (Simontwice) [inline research/conditional/utils/argparse.py:16]:
```suggestion
    )  # not used by granularity
```

**Review by Simontwice — APPROVED** [review]
LGTM, congrats
